### PR TITLE
Leave Scope Added

### PR DESCRIPTION
**Pull request by ImDineshSaini**
Leave scope added and now the function can be used as ModelName.leaves

### CODE REVIEW

**Comment by ImDineshSaini** [comment]
Leave option added - dependent function added

**Comment by antstorm** [comment]
@ImDineshSaini this PR is a total rip off of #246

**Comment by ImDineshSaini** [comment]
Yes and one more changes required I have seen the Rails 4 is work with mysql2 gem so  text_type = if ActiveRecord::Base.connection.adapter_name.downcase == 'mysql2' replaced by  text_type = if ActiveRecord::Base.connection.adapter_name.downcase == 'mysql'

**Comment by ImDineshSaini** [comment]
this PR is a total rip off of #246  + Fix the mysql connection error with rails 4 and utf issue.

**Comment by antstorm** [comment]
@ImDineshSaini all of that is fixed in #246 and tested

**Comment by ImDineshSaini** [comment]
I have tries with Rails 4.2.3 and mysql and it give me error and reason is that mysql2 is use as connection not mysql.

File: lib/ancestry/class_methods.rb [hunk 1]
```diff
@@ -17,6 +17,17 @@ def scope_depth depth_options, depth
       end
     end
 
+    # Scope that returns all the leaves
+    def leaves
+      id_column = "#{table_name}.id"
+      id_column_as_text = sql_cast_as_text(id_column)
+      parent_ancestry = sql_concat("#{table_name}.#{ancestry_column}", "'/'", id_column_as_text)
+
+      joins("LEFT JOIN #{table_name} AS c ON c.#{ancestry_column} = #{id_column_as_text} OR c.#{ancestry_column} = #{parent_ancestry}").
```
Comment (kbrock) [inline lib/ancestry/class_methods.rb]: is there a way to use an existing scope here?
Only adding the `group` and `having`

Also, should `id` be `primary_key`?

Comment (kbrock) [reply lib/ancestry/class_methods.rb]: better yet, is there a way to extract this concept so we can use it more?

File: lib/ancestry/class_methods.rb [hunk 2]
```diff
@@ -202,5 +208,26 @@ def rebuild_depth_cache!
...
+
+    def sql_cast_as_text column
+      #need mysql2 as new rails version support the mysql2 gem only
+      text_type = if ActiveRecord::Base.connection.adapter_name.downcase == 'mysql2'
+        'CHAR'
+      else
+        'TEXT'
+      end
+      # updated mysql thrown error if utf is not defined
+      "CAST(#{column} AS #{text_type}) collate utf8_unicode_ci"
```
Comment (kbrock) [inline lib/ancestry/class_methods.rb]: think the `collate` is better up with mysql only

Comment (antstorm) [reply lib/ancestry/class_methods.rb]: @kbrock why are you even looking at this PR? The original one is #246

Comment (kbrock) [reply lib/ancestry/class_methods.rb]: please be nice

**Comment by kbrock** [comment]
Closing in favor of #246

**Pull request closed without merging**